dev/AppNotifications/AppNotificationUtility.cpp

```diff
@@ -232,79 +234,124 @@ std::wstring Microsoft::Windows::AppNotifications::Helpers::SetDisplayNameBasedO
-
-        THROW_HR_IF_MSG(E_UNEXPECTED, propVariantIcon.vt != VT_LPWSTR, "Icon should be a valid Unicode string");
+        THROW_HR_IF(E_UNEXPECTED, propVariantIcon.vt != VT_LPWSTR || Microsoft::Windows::AppNotifications::Helpers::IsWideStringEmptyOrNull(propVariantIcon.pwszVal));
 
-        THROW_HR_IF_MSG(E_UNEXPECTED, wcslen(propVariantIcon.pwszVal) == 0, "Icon is an empty string");
+        std::wstring localIconFilePathAsWstring = propVariantIcon.pwszVal;
 
-        iconFilePath = propVariantIcon.pwszVal;
-
-        // Icon filepaths from Shell APIs have this format: <filepath>,-<index>,
-        // since .ico files can have multiple icons in the same file.
+        // Icon filepaths from Shell APIs usually follow this format: <filepath>,-<index>,
+        // since .ico or .dll files can have multiple icons in the same file.
         // NotificationController doesn't seem to support such format, so let it take the first icon by default.
-        auto iteratorForCommaDelimiter{ iconFilePath.find_first_of(L",") };
+        auto iteratorForCommaDelimiter{ localIconFilePathAsWstring.find_first_of(L",") };
         if (iteratorForCommaDelimiter != std::wstring::npos) // It may or may not have an index, which is fine.
         {
-            iconFilePath.erase(iteratorForCommaDelimiter);
+            localIconFilePath = wil::make_unique_string<wil::unique_cotaskmem_string>(localIconFilePathAsWstring.erase(iteratorForCommaDelimiter).c_str());
         }
 
-        auto iteratorForFileExtension{ iconFilePath.find_first_of(L".") };
+        *displayName = localDisplayName.release();
+        *iconFilePath = localIconFilePath.release();
+
+        return S_OK;
+    }
+
+    return HRESULT_FROM_WIN32(ERROR_NOT_FOUND);
+}
+
+void Microsoft::Windows::AppNotifications::Helpers::RegisterAssets(std::wstring const& appId, std::wstring const& clsid)
+{
+    wil::unique_hkey hKey;
+    // subKey: \Software\Classes\AppUserModelId\{AppGUID}
+    std::wstring subKey{ c_appIdentifierPath + appId };
+
+    THROW_IF_WIN32_ERROR(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr /* lpClass */,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr /* lpSecurityAttributes */,
+        &hKey,
+        nullptr /* lpdwDisposition */));
+
+    bool useDefaultAssets{ false };
+
+    wil::unique_cotaskmem_string displayName;
+    wil::unique_cotaskmem_string iconFilePath;
+
+    // Try the following techniques to retrieve display name and icon:
+    // 1. From the current process.
+    // 2. Based on the best app shortcut, using the FrameworkUdk.
+    // 3. From the foreground window.
+    // 4. Use the default assets.
+    if (FAILED(RetrieveAssetsFromProcess(&displayName, &iconFilePath)) &&
+        FAILED(ToastNotifications_RetrieveAssets_Stub(&displayName, &iconFilePath)) &&
+        FAILED(RetrieveAssetsFromWindow(&displayName, &iconFilePath)))
```

Comment:
FAILED(...)? Not FAILED(LOG_IF_FAILED(...))?

Reply:
Is it to log the error? Also, does LOG_IF_FAILED return the corresponding HRESULT?